### ISSUE TITLE
.gitignore not being properly parsed

### ISSUE THREAD

**Issue by rileyrg** [ggreer/the_silver_searcher#1402]
If I use the "!" notation to reinclude subdirectories ag doesn't cope. 

eg my .gitignore

```
*
!/editor-config
!/editor-config/*

```

ag will not search editor-config. This is quite a major.

**Comment by JaapJoris** [comment]
Here's a "me too!" from me. I ran into this issue and found this page via google.

**Comment by vmenajr** [comment]
Just hit this today and though not ideal the `-u` works around the problem.  Small price to pay for an amazing tool.

**Comment by rileyrg** [comment]
> Just hit this today and though not ideal the `-u` works around the problem. Small price to pay for an amazing tool.

Yup it is / was. I moved onto ripgrep which parses globs properly in a hierarchical manner. And it's faster I think: certainly I had more success inside emacs with counsel-ripgrep over other implementations.

**Comment by systemmonkey42** [comment]
This is not the only problem with .gitignore.  I'm finding if I have a path starting with '/' (eg, '/node_modules') and I run 'ag' from the same directory as the .gitignore, it correctly ignore the node_modules folder.

However if I run the same query from the parent directory, it misinterprets the '/node_modules' and proceeds to scan the '/repo/node_modules' folder.

I would expect a path starting with a '/' in a .gitignore (or any ignore file) would implicitly mean `<subdir>/exclusion` where `<subdir>` is the path to the ignore file.

Does this make sense?

**Comment by marc-guenther** [comment]
Sorry for hijacking this, but for me, `.gitignore` doesn't work at all (see #1381), not even for a simple entry like `target`. I have to symlink it to `.ignore` to make it work.

Apparently for you it works without the link? Could you quickly confirm?
